### PR TITLE
SB-300: Updating Group Repository Index

### PR DESCRIPTION
**Pull request by fuss86**

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java [hunk 1]
```diff
@@ -203,20 +203,18 @@ public void deleteFromIndex(RepositoryPath path)
             throws IOException
     {
         Repository repository = path.getFileSystem().getRepository();
-        RepositoryFileAttributes a = (RepositoryFileAttributes) Files.readAttributes(path, BasicFileAttributes.class);
 
-        if (!repositoryFeatures.isIndexingEnabled(repository) || a.isMetadata())
+        if (!repositoryFeatures.isIndexingEnabled(repository) || isMetadata(path.toString()))
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: why is this changed? we have common way to do it thru `RepositoryFileAttributes`.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java [hunk 2]
```diff
@@ -203,20 +203,18 @@ public void deleteFromIndex(RepositoryPath path)
...
 
         final RepositoryIndexer indexer = getRepositoryIndexer(path);
         if (indexer != null)
         {
-            String extension = path.getFileName().toString().substring(
-                    path.getFileName().toString().lastIndexOf('.') + 1);
+            String extension = FilenameUtils.getExtension(path.getFileName().toString());
 
-            MavenArtifactCoordinates coordinates = (MavenArtifactCoordinates) a.getCoordinates();
+            MavenArtifactCoordinates coordinates = getArtifactCoordinates(path.toString());
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: same question as above

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                                                                        groupRepository.getId(),
+                                                                        artifactRelativePath,
+                                                                        repositoryArtifactExistence);
+        }
+    }
+
+    /**
+     * This method goes down the group repository tree and lookups for an existence of the artifact in the given path.
+     */
+    private void checkArtifactExistenceInTheGroupRepositorySubtree(final Repository groupRepository,
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java]: maybe this can be implemented generic way? I mean that the logic is pretty generic but class called `BaseMavenGroupRepositoryComponent`

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java (hunk 1 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: Because 

`RepositoryFileAttributes a = (RepositoryFileAttributes) Files.readAttributes(path, BasicFileAttributes.class);`

requires physicall existence of the file, but in terms of group repositories there won't be any artifacts.

`deleteFromIndex` should also work for group repositories

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java (hunk 3 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java]: Thanks. Definietely it requires some code reorganization. It's still in progress. Thanks again.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: right,
thanks for the explanation
maybe `RepositoryFileAttributes` should work for non-existent files too, I will look at this in future

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/group/GroupRepositoryArtifactExistenceChecker.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    private LayoutProviderRegistry layoutProviderRegistry;
+
+    public boolean artifactExistsInTheGroupRepositorySubtree(final Repository groupRepository,
+                                                             final String artifactPath)
+            throws IOException
+    {
+        return artifactExistsInTheGroupRepositorySubtree(groupRepository, artifactPath, new HashMap<>());
+    }
+
+    public boolean artifactExistsInTheGroupRepositorySubtree(final Repository groupRepository,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/group/GroupRepositoryArtifactExistenceChecker.java]: Cosmetic: I think there's a typo in the camel case - `artifactExistsInTheGroupRepositorySubTree`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ConfigurationManagementServiceImpl.java [hunk 5]
```diff
@@ -187,28 +187,30 @@ public Repository getRepository(String storageId,
     }
 
     @Override
-    public List<Repository> getGroupRepositoriesContaining(String repositoryId)
+    public List<Repository> getGroupRepositoriesContaining(String storageId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ConfigurationManagementServiceImpl.java]: Yeah, I think you're right... I initially did it like this because I was thinking about the case where multiple storages could have repositories with the same ID. 
I have the feeling that we'll probably still need the old method at some point as well, but let's leave it as it is.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/MavenArtifactStoredEventListener.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Inject
+    private MavenIndexGroupRepositoryComponent mavenIndexGroupRepositoryComponent;
+
+    @Override
+    public void handle(final ArtifactEvent event)
+    {
+        final Repository repository = getRepository(event);
+
+        if (!RepositoryLayoutEnum.MAVEN_2.getLayout().equals(repository.getLayout()))
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/MavenArtifactStoredEventListener.java]: Please, avoid using the `RepositoryLayoutEnum.MAVEN_2.getLayout()`, as I'm thinking of removing this (you can check #437 for more details). Where possible, could you please use `Maven2LayoutProvider.ALIAS` instead...? Thanks!

I know, it's going to be tricky to remove this enum because it's used in places like `Repository` for the default layout.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/MavenArtifactStoredEventListener.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        if (event.getType() != ArtifactEventTypeEnum.EVENT_ARTIFACT_FILE_STORED.getType())
+        {
+            return;
+        }
+
+        try
+        {
+            mavenIndexGroupRepositoryComponent.updateGroupsContaining(event.getStorageId(), event.getRepositoryId(),
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/MavenArtifactStoredEventListener.java]: Cosmetic: Please, keep chopped arguments on separate lines. Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java [hunk 8]
```diff
@@ -215,27 +197,22 @@ public void deleteFromIndex(RepositoryPath path)
...
-                                                                      coordinates.getGroupId(),
-                                                                      coordinates.getArtifactId(),
-                                                                      coordinates.getVersion(),
-                                                                      coordinates.getClassifier(),
-                                                                      extension)));
+                                                                      artifact.getGroupId(),
+                                                                      artifact.getArtifactId(),
+                                                                      artifact.getVersion(),
+                                                                      artifact.getClassifier(),
+                                                                      artifact.getType())));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: I am not entirely sure whose version is more correct. To get better perspective, would you mind telling me why you chose `artifact.getType()` here?
I have some distant memory that version 6.0 of the Maven Indexer was leaning towards dropping the use of `artifact.getType()`... (I may be wrong).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+            }
+
+            cleanupGroupsContaining(groupRepository.getStorage().getId(),
+                                    groupRepository.getId(),
+                                    artifactPath,
+                                    repositoryArtifactExistence);
+        }
+    }
+
+    protected abstract void cleanupGroupWhenArtifactPathNoLongerExistsInSubtree(Repository groupRepository,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java]: Cosmetic: Same as my earlier comment about the camel-hump typo.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        return false;
+    }
+
+    protected abstract UpdateCallback newInstance();
+
+    protected Repository getRepository(final String storageId,
+                                       final String repositoryId)
+    {
+        return configurationManagementService.getConfiguration().getStorage(
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java]: Cosmetic: Indentation - could be single line, or chopped.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/services/impl/ArtifactIndexesServiceImpl.java [hunk 11]
```diff
@@ -43,6 +51,34 @@
...
+        if (!features.isIndexingEnabled(repository) || indexer == null)
+        {
+            return;
+        }
+
+        String artifactRelativePath = RepositoryFiles.relativizeUri(artifactPath).toString();
+        Artifact artifact = ArtifactUtils.convertPathToArtifact(artifactRelativePath);
+
+        File storageBasedir = new File(storage.getBasedir());
+        File artifactFile = new File(new File(storageBasedir, repository.getId()),
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/services/impl/ArtifactIndexesServiceImpl.java]: Cosmetic: Indentation.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/group/GroupRepositoryArtifactExistenceChecker.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+public class GroupRepositoryArtifactExistenceChecker
+{
+
+    @Inject
+    private ConfigurationManager configurationManager;
+
+    @Inject
+    private LayoutProviderRegistry layoutProviderRegistry;
+
+    public boolean artifactExistsInTheGroupRepositorySubtree(final Repository groupRepository,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/group/GroupRepositoryArtifactExistenceChecker.java]: Cosmetic: I think there's a typo in the camel case - `artifactExistsInTheGroupRepositorySubTree`.

**Comment by carlspring** [comment]
Hi @fuss86,

I have a few questions:
* How does all this work for a newly created group repository?
* Do we have to create a single-run cron task that populates it once it's created?

Martin

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java (hunk 8 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: > I am not entirely sure whose version is more correct.

If you mean that I dropped previous implementation:
`RepositoryFileAttributes a = Files.readAttributes(path, RepositoryFileAttributes.class);` please note that this solution required physical file existence in the file system. In terms of group repositories, artifacts won't be existing physically.

> To get better perspective, would you mind telling me why you chose artifact.getType() here?

Please notice where did the `Artifact artifact` variable come from. It has been created from `org.carlspring.maven.commons.util.ArtifactUtils#convertPathToArtifact`. Please take a look at the `type` field:
https://github.com/carlspring/maven-commons/blob/master/src/main/java/org/carlspring/maven/commons/util/ArtifactUtils.java#L129

If you compare it with the previous implementation `String extension = path.getFileName().toString().substring(path.getFileName().toString().lastIndexOf('.') + 1);` you'll find out that it's logically the same.

**Comment by fuss86** [comment]
> How does all this work for a newly created group repository?

Currently, newly created group repository will only be updated based on the future activities in the sub-tree repositories. 

> Do we have to create a single-run cron task that populates it once it's created?

I'll prepare some component and trigger for that.

> Could we add some tests for the above?

Of course.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/group/metadata/MavenMetadataGroupRepositoryComponentTest.java [hunk 13]
```diff
@@ -37,101 +29,26 @@
 @RunWith(SpringJUnit4ClassRunner.class)
 @ContextConfiguration(classes = Maven2LayoutProviderTestConfig.class)
 public class MavenMetadataGroupRepositoryComponentTest
-        extends TestCaseWithMavenArtifactGenerationAndIndexing
+        extends BaseMavenGroupRepositoryComponentTest
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/group/metadata/MavenMetadataGroupRepositoryComponentTest.java]: @sbespalov : Might be worth noting this renaming.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/TestCaseWithMavenArtifactGeneration.java [hunk 14]
```diff
@@ -32,7 +31,7 @@
  * @author mtodorov
  */
 public class TestCaseWithMavenArtifactGeneration
-        extends TestCaseWithRepository
+        extends TestCaseWithRepositoryManagement
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/TestCaseWithMavenArtifactGeneration.java]: Yeah, this name makes more sense... :)
